components/sup/src/manager/service/config.rs

```diff
@@ -651,6 +655,138 @@ mod test {
         }
     }
 
+    struct TestPkg {
+        base_path: PathBuf,
+    }
+
+    impl TestPkg {
+        fn new(tmp: &TempDir) -> Self {
+            let pkg = Self { base_path: tmp.path().to_owned() };
+
+            fs::create_dir_all(pkg.default_config_dir()).expect(
+                "create deprecated user config dir",
+            );
+            fs::create_dir_all(pkg.recommended_user_config_dir())
+                .expect("create recommended user config dir");
+            fs::create_dir_all(pkg.deprecated_user_config_dir())
+                .expect("create default config dir");
+            pkg
+        }
+
+        fn deprecated_ucp(&self) -> PathBuf {
+            self.deprecated_user_config_dir().join("user.toml")
+        }
+
+        fn recommended_ucp(&self) -> PathBuf {
+            self.recommended_user_config_dir().join("user.toml")
+        }
+    }
+
+    impl PackageConfigPaths for TestPkg {
+        fn name(&self) -> String {
+            String::from("testing")
+        }
+        fn default_config_dir(&self) -> PathBuf {
+            self.base_path.join("root")
+        }
+        fn recommended_user_config_dir(&self) -> PathBuf {
+            self.base_path.join("user")
+        }
+        fn deprecated_user_config_dir(&self) -> PathBuf {
+            self.base_path.join("svc")
+        }
+    }
+
+    struct TestToml {
+        text: String,
+        value: toml::Value,
+    }
+
+    impl TestToml {
+        fn new(text: &str) -> Self {
+            Self {
+                text: String::from(text),
+                value: Self::parse(text),
+            }
+        }
+
+        fn parse(text: &str) -> toml::Value {
+            toml::Value::Table(toml_from_str(text))
+        }
+
+        fn write(&self, file: &mut File) {
+            file.seek(SeekFrom::Start(0)).expect(
+                "seek in toml file to start",
+            );
+            file.set_len(0).expect("truncate toml file");
+            file.write_all(self.text.as_bytes()).expect(
+                "write raw toml value",
+            );
+            file.flush().expect("flush changes in toml file");
+
+            file.seek(SeekFrom::Start(0)).expect(
+                "seek in toml file to start for checking",
+            );
+            let mut buf = String::new();
+            file.read_to_string(&mut buf).expect(
+                "read value from toml file",
+            );
+            assert_eq!(buf, self.text);
```

Comment:
I think this implementation is more complex than it needs to be for these tests; you probably just need a `write_all` and a `flush`. The `assert_eq!` is basically testing that Rust's I/O code is working.

Beyond that, though, I think there's a larger opportunity to simplify this test code. There's a lot of repetition in the test cases themselves around opening a file, creating a representation of the desired TOML, and then writing it. In the end, you just want one function that takes a path and some content, and writes that data to the right place. That one function could hide all the details of file opening, formatting, and writing, which are really incidental to the tests themselves. When you're reading the tests, you just want one line that says "OK, I'm creating this config file at this location"; you don't want to worry about the details of how that's done.

Reply:
I added this assert_eq, because earlier I had a `reset` function that cleared a file and wrote new contents there. So I truncated it and then wrote new contents only to find out that I got a bunch of zeros at the beginning of the file, because truncating is not seeking.

Anyway, I'll have a look how to make it simpler.